core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -4907,15 +4938,34 @@ public void rebuildDependencyGraph() {
      * @since 1.522
      */
     public Future<DependencyGraph> rebuildDependencyGraphAsync() {
-        dependencyGraphDirty.set(true);
+        synchronized (dependencyGraphLock) {
+            // Collect calls to this method to avoid unnecessary calculation of the dependency graph
+            if (scheduledFutureDependencyGraph != null) {
+                return scheduledFutureDependencyGraph;
+            }
+            // Schedule new calculation
+            return scheduledFutureDependencyGraph = scheduleCalculationOfFutureDependencyGraph(500, TimeUnit.MILLISECONDS);
+        }
+    }
+
+
+    private Future<DependencyGraph> scheduleCalculationOfFutureDependencyGraph(int delay, TimeUnit unit) {
         return Timer.get().schedule(() -> {
-            if (dependencyGraphDirty.get()) {
-                rebuildDependencyGraph();
+            synchronized (dependencyGraphLock) {
+                //Wait for the currently running calculation to finish
+                calculatingFutureDependencyGraph.get();
```

Comment:
You might not want to lock while waiting for this to finish, I don't know how long the graph build actually takes but for large instances this may be non-trivial time

Reply:
Fixed